integration-test/src/test/scala/com/cloudera/livy/test/BatchIT.scala

```diff
@@ -99,50 +182,87 @@ class BatchIT extends BaseIntegrationTestSuite {
     val request = new CreateBatchRequest()
     request.file = script
     request.args = args
-    runBatch(request)
+    waitAndDeleteBatch(startBatch(request).id)
   }
 
-  private def runSpark(klass: Class[_], args: List[String] = Nil): SessionInfo = {
+  private def runSpark(
+      klass: Class[_],
+      args: List[String] = Nil,
+      waitForExit: Boolean = true): SessionInfo = {
     val request = new CreateBatchRequest()
     request.file = testLibPath
-    request.className = Some(klass.getName())
+    request.className = Some(klass.getName)
     request.args = args
-    runBatch(request)
+
+    val batchId = startBatch(request)
+    if (waitForExit) waitAndDeleteBatch(batchId.id) else batchId
   }
 
-  private def runBatch(request: CreateBatchRequest): SessionInfo = {
+  private def getBatchSessionInfo(batchId: Int): SessionInfo = {
+    val r = httpClient.prepareGet(s"$livyEndpoint/batches/$batchId")
+      .execute()
+      .get()
+    assert(r.getStatusCode() === SC_OK)
+    mapper.readValue(r.getResponseBodyAsStream(), classOf[SessionInfo])
+  }
+
+  private def deleteBatch(batchId: Int): Unit = {
+    val r = httpClient.prepareDelete(s"$livyEndpoint/batches/$batchId")
+      .execute()
+      .get()
+    assert(r.getStatusCode() === SC_OK)
+  }
+
+  private def startBatch(request: CreateBatchRequest): SessionInfo = {
     request.conf = Map("spark.yarn.maxAppAttempts" -> "1")
 
     val response = httpClient.preparePost(s"$livyEndpoint/batches")
       .setBody(mapper.writeValueAsString(request))
       .execute()
       .get()
 
-    assert(response.getStatusCode() === SC_CREATED)
+    withClue(response) {
+      assert(response.getStatusCode === SC_CREATED)
+    }
+
+    mapper.readValue(response.getResponseBodyAsStream(), classOf[SessionInfo])
+  }
+
+  private def waitBatchForRunning(batchId: Int): ApplicationId = {
+    eventually {
+      val batch = getBatchSessionInfo(batchId)
 
-    val batchInfo = mapper.readValue(response.getResponseBodyAsStream(), classOf[SessionInfo])
+      // If batch state transits to any error state, fail test immediately.
+      if (batch.state == SessionState.Error().toString() ||
+        batch.state == SessionState.Dead().toString()) {
+        throw new TestPendingException {
```

Comment:
I see this is because of using `eventually` above. Anyway this will affect the reporting for this test, since per the javadoc it doesn't seem like it would show up as a proper test failure.